### ISSUE TITLE
Support work tag overload final() function in parallel dispatch functor 

### ISSUE THREAD

**Issue by Char-Aznable** [kokkos/kokkos#3651]
Currently functor used for parallel dispatch can't have a work-tagged final() function, in contrast to the init() and join(), which can be work-tagged

**Comment by crtrott** [comment]
I think the problem is largely here:
https://github.com/kokkos/kokkos/blob/734f577adb588e5ebef8760ca36c7815312accff/core/src/impl/Kokkos_FunctorAdapter.hpp#L2041-L2079

We need a test for this too.

**Comment by dalg24** [comment]
@Char-Aznable This has been fixed and merged into develop.

**Comment by Char-Aznable** [comment]
@dalg24 @masterleinad Actually, I think it's still missing some cases here:
```c++
#include <limits>
#include <Kokkos_Core.hpp>

using ExecutionSpace = Kokkos::DefaultExecutionSpace;

struct TagReduce {};

template < class ViewMatrix, class ViewVector >
struct ColumnwiseMax {

  using V = typename ViewMatrix::value_type;
  using size_type = typename ViewMatrix::size_type;

  using value_type = V[];

  KOKKOS_INLINE_FUNCTION
  void operator()(const TagReduce, const size_type iRow, value_type update) const {
    for(size_type jCol = 0; jCol < value_count; ++jCol) {
      if(data_(iRow, jCol) > update[jCol]) {
        update[jCol] = data_(iRow, jCol);
      }
    }
  }

  KOKKOS_INLINE_FUNCTION void join(const TagReduce, volatile value_type dst,
    const volatile value_type src) const {
    for(size_type i = 0; i < value_count; ++i) {
      if(src[i] > dst[i]) { dst[i] = src[i]; }
    }
  }

  KOKKOS_INLINE_FUNCTION void init(const TagReduce, value_type dst) const {
    for(size_type i = 0; i < value_count; ++i) { dst[i] = -std::numeric_limits<V>::infinity(); }
  }

  using has_final = std::true_type;
  KOKKOS_INLINE_FUNCTION void final(const TagReduce, value_type src) const {
    for(size_type i = 0; i < value_count; ++i) { results_(i) = src[i]; }
  }

  ViewMatrix data_;
  ViewVector results_;
  size_type value_count;

};

int main(int argc, char* argv[]) {
  Kokkos::initialize(argc,argv); {

  const size_t numRows = 4;
  const size_t numCols = 2374;
  using ViewMatrix = Kokkos::View<float**>;
  using ViewVector = Kokkos::View<float*>;

  ViewMatrix m ("m", numRows, numCols);
  ViewVector results("results", numCols);
  Kokkos::deep_copy(m, 1.0);
  // ... fill X before the following ...
  Kokkos::parallel_reduce (Kokkos::RangePolicy<ExecutionSpace,TagReduce>(0, numRows),
    ColumnwiseMax<ViewMatrix,ViewVector>{m, results, numCols});

  Kokkos::fence();

  } Kokkos::finalize();
}

```
gives the compilation error:
```c++
/home/aznb/mycodes/testKokkos/build_clang_debug/_deps/kokkos-src/core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:313:23: error: ambiguous partial specializations of 'FunctorFinal<ColumnwiseMax<Kokkos::View<float **>, Kokkos::View<float *>>, TagReduce, float *, void>'
        Kokkos::Impl::FunctorFinal<ReducerTypeFwd, WorkTagFwd>::final(
                      ^
/home/aznb/mycodes/testKokkos/build_clang_debug/_deps/kokkos-src/core/src/Kokkos_Parallel_Reduce.hpp:877:13: note: in instantiation of member function 'Kokkos::Impl::ParallelReduce<ColumnwiseMax<Kokkos::View<float **>, Kokkos::View<float *>>, Kokkos::RangePolicy<Kokkos::OpenMP, TagReduce>, Kokkos::InvalidType, Kokkos::OpenMP>::execute' requested here
    closure.execute();
            ^
/home/aznb/mycodes/testKokkos/build_clang_debug/_deps/kokkos-src/core/src/Kokkos_Parallel_Reduce.hpp:1153:50: note: in instantiation of member function 'Kokkos::Impl::ParallelReduceAdaptor<Kokkos::RangePolicy<Kokkos::OpenMP, TagReduce>, ColumnwiseMax<Kokkos::View<float **>, Kokkos::View<float *>>, Kokkos::View<float *, Kokkos::HostSpace, Kokkos::MemoryTraits<1>>>::execute' requested here
                              result_view_type>::execute("", policy, functor,
                                                 ^
/home/aznb/mycodes/testKokkos/testReduceArray.cpp:67:11: note: in instantiation of function template specialization 'Kokkos::parallel_reduce<Kokkos::RangePolicy<Kokkos::OpenMP, TagReduce>, ColumnwiseMax<Kokkos::View<float **>, Kokkos::View<float *>>>' requested here
  Kokkos::parallel_reduce (Kokkos::RangePolicy<ExecutionSpace,TagReduce>(0, numRows),
          ^
/home/aznb/mycodes/testKokkos/build_clang_debug/_deps/kokkos-src/core/src/impl/Kokkos_FunctorAdapter.hpp:2136:8: note: partial specialization matches [with FunctorType = ColumnwiseMax<Kokkos::View<float **>, Kokkos::View<float *>>, ArgTag = TagReduce, T = float]
struct FunctorFinal<
       ^
/home/aznb/mycodes/testKokkos/build_clang_debug/_deps/kokkos-src/core/src/impl/Kokkos_FunctorAdapter.hpp:2153:8: note: partial specialization matches [with FunctorType = ColumnwiseMax<Kokkos::View<float **>, Kokkos::View<float *>>, ArgTag = TagReduce, T = float]
struct FunctorFinal<FunctorType, ArgTag,
       ^
In file included from /home/aznb/mycodes/testKokkos/testReduceArray.cpp:10:
In file included from /home/aznb/mycodes/testKokkos/build_clang_debug/_deps/kokkos-src/core/src/Kokkos_Core.hpp:56:
In file included from /home/aznb/mycodes/testKokkos/build_clang_debug/_deps/kokkos-build/KokkosCore_Config_DeclareBackend.hpp:47:
In file included from /home/aznb/mycodes/testKokkos/build_clang_debug/_deps/kokkos-src/core/src/decl/Kokkos_Declare_OPENMP.hpp:49:
In file included from /home/aznb/mycodes/testKokkos/build_clang_debug/_deps/kokkos-src/core/src/Kokkos_OpenMP.hpp:227:
/home/aznb/mycodes/testKokkos/build_clang_debug/_deps/kokkos-src/core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:313:63: error: incomplete definition of type 'Kokkos::Impl::FunctorFinal<ColumnwiseMax<Kokkos::View<float **>, Kokkos::View<float *>>, TagReduce, float *, void>'
        Kokkos::Impl::FunctorFinal<ReducerTypeFwd, WorkTagFwd>::final(
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
/home/aznb/mycodes/testKokkos/build_clang_debug/_deps/kokkos-src/core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:375:59: error: incomplete definition of type 'Kokkos::Impl::FunctorFinal<ColumnwiseMax<Kokkos::View<float **>, Kokkos::View<float *>>, TagReduce, float *, void>'
    Kokkos::Impl::FunctorFinal<ReducerTypeFwd, WorkTagFwd>::final(
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
3 errors generated.
CMakeFiles/testReduceArray.dir/build.make:81: recipe for target 'CMakeFiles/testReduceArray.dir/testReduceArray.cpp.o' failed
make[3]: *** [CMakeFiles/testReduceArray.dir/testReduceArray.cpp.o] Error 1
CMakeFiles/Makefile2:545: recipe for target 'CMakeFiles/testReduceArray.dir/all' failed
make[2]: *** [CMakeFiles/testReduceArray.dir/all] Error 2
CMakeFiles/Makefile2:552: recipe for target 'CMakeFiles/testReduceArray.dir/rule' failed
make[1]: *** [CMakeFiles/testReduceArray.dir/rule] Error 2
Makefile:196: recipe for target 'testReduceArray' failed
make: *** [testReduceArray] Error 2

```

**Comment by Char-Aznable** [comment]
xref #3897 would it be possible to make the required nested definition of `value_type`, `size_type`, `value_count` and `has_final` all templated by the tag?
